### DISCUSSION TITLE
bug/vulnerability in so-user?

### DISCUSSION THREAD

**Discussion by abesinger** [Security-Onion-Solutions/securityonion#4452] [Unsupported Versions] [answered]
After spending the day trying to diagnose why I couldn't log into FleetDM, I realized that it's because there are spaces in my password; password works for logging into SecurityOnion, but I can only login to FleetDM if I use the first word in my password.  Is that what was fixed in [#4249](https://github.com/Security-Onion-Solutions/securityonion/issues/4249#issue-896686295)? If so, it doesn't appear that the fix was propagated via soup; none of `/usr/sbin/so-` has been modified since then, and I ran `soup` yesterday.

I have some additional observations.  It seems that `so-user `will create a new account for FleetDM and TheHive, but password changing with `so-user` doesn't change passwords for either of those, so you're stuck with the password you initially chose. Now you can change passwords inside of FleetDM, but it has different password rules than SO, which is frankly annoying.

I think that password changing could be easily accomodated by changing `so-fleet-user-add` so that the mysql command does an insert-or-replace action (there's a couple of ways to do it), instead of just an insert. Or a` so-fleet-user-pw` script could be easily put together that does a mysql replace.

Furthermore, if the version of `so-user` that I have installed is current, and it's getting invoked via a web executable, then it's got a shell metacharacter vulnerability ; the password isn't quoted on lines  283-284:

```
    check_container thehive && echo $password | so-thehive-user-add "$email"
    check_container fleet && echo $password | so-fleet-user-add "$email"
```
If it's not current, then why isn't it getting updated via `soup`?

FYI, my install is 2.3.51

**Comment by dougburks** [comment, marked as answer]
Thanks for reporting this bug. We'll get the missing quotes fixed in the next release. You mentioned _vulnerability_ a couple times in this post, however the script will only run if the user is root or has sudo privileges. Is there another attack vector that we should consider?
Regarding the synchronization of passwords, this is an area we're interested in improving.

**Reply by abesinger** [reply]
Regarding vulnerability, I was thinking primarily of the possibility that the script is getting invoked via the web application -- i.e. when changing a password through the SO GUI.  If that's not the case, then that's less of a worry.

However, I can also point to a more obscure/rare possibility: Consider an environment where a sysadmin gives a user, say, a help desk person, limited sudo privileges -- in this case only the ability to run so-user.  The user (or attacker who compromised the account) could then craft a password that would give them a root shell, bypassing their sudo restrictions. 

It's an obscure example, but sometimes the rare/overlooked thing is what you get screwed by.  And the fix is pretty simple, so might as well fix it.

**Reply by dougburks** [reply]
Yep, we've already created https://github.com/Security-Onion-Solutions/securityonion/issues/4460 for this.

**Reply by abesinger** [reply]
Regarding the synchronizing of passwords, it's really a pretty easy fix.  For FleetDM, you could simply do it by copying `so-fleet-user-add` to `so-fleet-user-pw` and changing line 55 to:

```
MYSQL_OUTPUT=$(docker exec so-mysql mysql -u root --password=$MYSQL_PASS fleet -e \
    "UPDATE users SET password='$FLEET_HASH', salt='') where username='$FLEET_USER'" 2>&1)
```

I haven't actually tested that because I don't want to risk screwing up my server, caveat coder .  You'd also have to add a line in `so-user` to invoke the script, I'll leave that one as an exercise for the reader.

Looking at `so-thehive-user-add`, changing TheHive passwords is a little less obvious, but may be about as simple.
